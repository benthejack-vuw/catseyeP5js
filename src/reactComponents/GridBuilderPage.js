import React from 'react'

const GridBuilderPage = () => (
    <div>
        GRID BUILDER    
    </div>
)

export default GridBuilderPage;