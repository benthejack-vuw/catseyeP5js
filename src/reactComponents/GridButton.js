import React from 'react'

import store from '../store/configureStore'


const GridButton = (props) => (
    <div class='grid-button'>
    </div>
)

export default GridButton;