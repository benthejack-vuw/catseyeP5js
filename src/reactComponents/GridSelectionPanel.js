import React from 'react'

import store from '../store/configureStore'


const GridSelectionPanel = (props) => (
    <div>
        {
            props.grids &&
            props.grids.map((grid)=>(
                <GridButton key={grid.name} {...grid} />
            ))
        }
    </div>
)

export default GridSelectionPanel;