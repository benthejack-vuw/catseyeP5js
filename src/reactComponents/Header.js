import React from 'react';
import { Link } from 'react-router-dom'

export const Header = ({startLogout}) => (  
    <div className='header'>
        <div className='container'>
            <header className='header__layout'>
                <Link className="header__home-link" to="/">
                    <h1>Catseye</h1>
                </Link>
                
                <Link className="header__help-link" to="/help">
                    <h2>Help</h2>
                </Link>
            </header>
        </div>
    </div>
);



export default Header;