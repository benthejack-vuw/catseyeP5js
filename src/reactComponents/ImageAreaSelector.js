import React from 'react'

export default class ImageAreaSelector extends React.Component{

	constructor(props){
        super(props);

        this.state = {
            size:props.size,
            coords: [{x:0, y:0}, {x:props.size.x, y:0}, {x:props.size.x, y:props.size.y}],
            selected: undefined
        }

        document.addEventListener('mouseup', this.stopDragging);
    }

    componentDidMount = () => {
        this.centerImage();
        this.constrainSelectorsToImage();
    }

    componentWillUnmount = () => {
        document.removeEventListener('mouseup', this.stopDragging);
    }

    updateMouseCoords = (e) => {
        const rect = document.getElementById(`${this.idPrefix()}-container`).getBoundingClientRect();
        const x = e.clientX - rect.left; 
        const y = e.clientY - rect.top;

        this.setState((prev)=>{
            const coords = [...prev.coords];
            if(prev.selected !== undefined){
                coords[prev.selected] = {x,y};
            }
            return {coords};
        });

        this.constrainSelectorsToImage();
        this.props.updateCallback(this.uvCoordinates());
    }

    setFromTextureCoordinates = (textureCoordinates) => {
        const bounds = this.imageBounds();

        const coords = textureCoordinates.map((coord)=>{
            return {
                x:coord.x*bounds.width + bounds.left,
                y:coord.y*bounds.height + bounds.top
            }
        });

        this.setState(()=>({ coords }))
    }

    uvCoordinates = () => {
        const bounds = this.imageBounds();

        return this.state.coords.map((coord)=>{
            return {
                x:(coord.x - bounds.left)/bounds.width,
                y:(coord.y - bounds.top)/bounds.height
            }
        });
    }
    
    imageBounds = () => {
        const containerRect = document.getElementById(`${this.idPrefix()}-container`).getBoundingClientRect();
        const imageRect = document.getElementById(`${this.idPrefix()}-selecton-image`).getBoundingClientRect();
        return new DOMRect(imageRect.x - containerRect.x, imageRect.y - containerRect.y, imageRect.width, imageRect.height);
    }

    centerImage = () => {
        const imageElem = document.getElementById(`${this.idPrefix()}-selecton-image`);
        const imageRect = imageElem.getBoundingClientRect();
        imageElem.setAttribute('x', (this.state.size.x-imageRect.width)/2 );
        imageElem.setAttribute('y', (this.state.size.y-imageRect.height)/2 );
    }

    constrainSelectorsToImage = () => {
        const bounds = this.imageBounds();
        this.setState(({coords}) => ({
            coords: coords.map(({x, y}) => ({
                x: Math.min(Math.max(bounds.left, x), bounds.right),
                y: Math.min(Math.max(bounds.top, y), bounds.bottom),
            }))
        }))
    }

    idPrefix = () => this.props.id || 'image-area-selector'

    startDragging = (index) => {
        this.setState( ({selected:prevSelected}) => ({selected: !prevSelected ? index : prevSelected}) ) 
    }

    stopDragging = () => {
        this.setState( () => ({selected: undefined}) )
    }

    dashedLines = () => {
        return this.state.coords.map((pt, index, array) => (
            <line 
                key={`selectionLine${index}`}
                strokeDasharray="5, 5"
                x1={pt.x}
                y1={pt.y}
                x2={array[(index+1)%array.length].x}
                y2={array[(index+1)%array.length].y}
                stroke='black'
            />
        ))
    }

    selectionCircles = () => {
        return this.state.coords.map((pt, index) => (
            <circle 
                key={`selection${index}`}
                className='image-selection-circle'                            
                cx={pt.x}
                cy={pt.y}
                r={20}
                onMouseDown = {()=>{this.startDragging(index)}}
            />
        ))
    }

    render = () => ( 
        <div>
            <svg id={`${this.idPrefix()}-container`}
             width={this.state.size.x}
             height={this.state.size.y}
             onMouseMove={this.updateMouseCoords}
            >
                <image 
                    id={`${this.idPrefix()}-selecton-image`}
                    xlinkHref={this.props.image}
                    width='100%'
                    onLoad={
                        (() => {
                            this.centerImage()
                            this.constrainSelectorsToImage()
                            this.setFromTextureCoordinates(this.props.textureCoordinates)
                        })
                    }
                />

                {this.dashedLines()}
                {this.selectionCircles()}
            </svg>
        </div>
    )

}
