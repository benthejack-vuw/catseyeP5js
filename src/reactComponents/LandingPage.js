import React from 'react'
import { Link } from 'react-router-dom'

const LandingPage = () => (
    <div>

        <Link to='/pattern-builder/'>
            <p>Build a pattern</p>
        </Link>

        <Link to='/grid-builder/'>
            <p>Build a grid</p>
        </Link>
    
    </div>
)

export default LandingPage;