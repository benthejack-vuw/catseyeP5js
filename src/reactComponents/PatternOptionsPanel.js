import React from 'react'
import { DomUtils } from 'bj-utils'
import { connect } from 'react-redux'
import { updateTextureCoordinates, updateImage, updateScale } from '../actions/PatternBuilder'
import ImageAreaSelector from './ImageAreaSelector'



export class PatternOptionsPanel extends React.Component{

    updateTexCoords = (coords)=>{
        this.props.updateTexCoords(coords);
    }
    
    loadImage = ()=>{
        DomUtils.selectImage().then((file)=>{
            return DomUtils.readImageAsURL(file);
        }).then((imageContentURL)=>{
            this.props.updateImage(imageContentURL);
        });
    }

    updateScale = (e)=>{
        this.props.updateScale(e.target.value);
    }

    render = ()=>(
        <div>
            <button onClick={this.loadImage}>Load Image</button>
            
            <ImageAreaSelector 
                image={this.props.image}
                textureCoordinates={this.props.textureCoordinates}
                size={{x:400,y:400}}
                updateCallback={this.updateTexCoords}
            />

            <input 
                type='range'
                min={0}
                max={2}
                step={0.01}
                value={this.props.scale}
                onChange={this.updateScale} 
            />
        </div>
    )
}


const mapStateToProps = (state) => ({
    image: state.patternBuilder.image,
    scale: state.patternBuilder.scale,
    textureCoordinates: state.patternBuilder.textureCoordinates,
});

const mapDispatchToProps = (dispatch) => ({
    updateImage: (image) => dispatch(updateImage(image)),
    updateScale: (scale) => dispatch(updateScale(scale)),
    updateTexCoords: (coords) => dispatch(updateTextureCoordinates(coords)),
});

export default connect(mapStateToProps, mapDispatchToProps) (PatternOptionsPanel);

