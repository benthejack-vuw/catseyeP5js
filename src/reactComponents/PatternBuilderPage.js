import React from 'react'

import PatternOptionsPanel from './PatternOptionsPanel'
import GridSelectionPanel  from './GridSelectionPanel'
import PatternDisplay from './PatternDisplay.js'

const PatternBuilderPage = () => (
    <div>
        <PatternDisplay/>
        <PatternOptionsPanel/>
        <GridSelectionPanel/>
    </div>
)

export default PatternBuilderPage;