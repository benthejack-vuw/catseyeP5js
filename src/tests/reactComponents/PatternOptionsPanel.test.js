import React from 'react';
import { shallow } from 'enzyme';
import { PatternOptionsPanel } from '../../reactComponents/PatternOptionsPanel'

import { defaults } from '../fixtures/PatternBuilderSettings'

test('should render PatternOptionsPanel correctly', ()=>{
    const wrapper = shallow(<PatternOptionsPanel 
                                image={defaults.image}
                                textureCoordinates={defaults.textureCoordinates}
                            />);
    expect(wrapper).toMatchSnapshot();
})

/*test('should call updateImage prop when the loadImage button was pressed', ()=>{
    
    const updateImageSpy = jest.fn();
    const wrapper = shallow(<PatternOptionsPanel 
                                image={defaults.image}
                                textureCoordinates={defaults.textureCoordinates}
                                updateImage={updateImageSpy}
                            />);
    wrapper.find('button').simulate('click', {});
    expect(updateImageSpy).toHaveBeenCalled();
})*/

test('should call updateScale prop when the scale slider is changed', ()=>{
    const updateScaleSpy = jest.fn();
    const wrapper = shallow(<PatternOptionsPanel 
                                image={defaults.image}
                                textureCoordinates={defaults.textureCoordinates}
                                updateScale={updateScaleSpy}
                            />);
    wrapper.find('input').simulate('change', {target:{value:0.12345}});
    expect(updateScaleSpy).toHaveBeenCalledWith(0.12345);
})