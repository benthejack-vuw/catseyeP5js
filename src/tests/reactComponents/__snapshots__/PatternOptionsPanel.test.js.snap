// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PatternOptionsPanel correctly 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Load Image
  </button>
  <ImageAreaSelector
    image="data:image/gif;base64,R0lGODlhEAAQAMQAAORHHOVSKudfOulrSOp3WOyDZu6QdvCchPGolfO0o/XBs/fNwfjZ0frl3/zy7////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAkAABAALAAAAAAQABAAAAVVICSOZGlCQAosJ6mu7fiyZeKqNKToQGDsM8hBADgUXoGAiqhSvp5QAnQKGIgUhwFUYLCVDFCrKUE1lBavAViFIDlTImbKC5Gm2hB0SlBCBMQiB0UjIQA7"
    size={
      Object {
        "x": 400,
        "y": 400,
      }
    }
    textureCoordinates={
      Array [
        Object {
          "x": 0,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 1,
        },
      ]
    }
    updateCallback={[Function]}
  />
  <input
    max={2}
    min={0}
    onChange={[Function]}
    step={0.01}
    type="range"
  />
</div>
`;
